Fix #47 by comparing `If-Modified-Since` and `stats.mtime` at same precision

jankeromnes opened the pull request:
Putting the drive-by fix from #48 here.

Addressed https://github.com/espadrine/sc/pull/48#discussion_r83327254

The solution works by applying the same second-truncation that browsers do. This means we suffer from the same caveats browsers do (intra-second file changes, or multiple changes right after a leap second, are edge cases), but every 304 implementation out there does too. I believe it's better to have a "working" 304 implementation with extremely rare caveats, rather than the current 304 implementation that compares second- with millisecond-precision and only works with a probability of 1/1000.

jankeromnes commented:
@espadrine Please take another look. I believe the most precise implementation one can build today.

espadrine commented on lib/camp.js:
```diff
@@ -629,13 +629,16 @@ function respondWithFile(req, res, path, ifNoFile) {
-      res.statusCode = 304;
-      res.end();  // not modified.
+    // Cache management (compare timestamps at second-level precision).
+    var modified = new Date(stats.mtime);
```
`stats.mtime` is already a Date.

Also, could you call it `lastModified`? It makes the if condition clearer.

jankeromnes replied:
Sure! However, the `new Date(stats.mtime)` is there to make a copy of the `stats.mtime` Date object that we can modify (we could do `stats.mtime.setMilliseconds(0)` directly, but that would be less clean).

espadrine commented on lib/camp.js:
```diff
@@ -629,13 +629,16 @@ function respondWithFile(req, res, path, ifNoFile) {
-      res.end();  // not modified.
+    // Cache management (compare timestamps at second-level precision).
+    var modified = new Date(stats.mtime);
+    modified.setMilliseconds(0);
```
I would recommend performing integer operations rather than Date operations, given the tremendous number of caveats JS Dates have as a consequence of their Java inheritance.

jankeromnes replied:
Ok, I'll rewrite this as an integer comparison between numbers of seconds.

espadrine commented on lib/camp.js:
```diff
@@ -629,13 +629,16 @@ function respondWithFile(req, res, path, ifNoFile) {
+    var modified = new Date(stats.mtime);
+    modified.setMilliseconds(0);
+    var since = req.headers['if-modified-since'];
+    if (since && (modified <= new Date(since))) {
```
… especially considering that this does type conversion to integers anyway.

jankeromnes commented:
Thanks for the review! Will update the documentation and push amended commits here soon.

jankeromnes commented:
@espadrine I rebased this pull request and addressed your comments: Comparison is now using integers instead of dates, and I've added a note about the 1/2-second caveat in the documentation. Please have another look.

jankeromnes commented:
:+1:
